File: Android/app/src/main/java/net/lonelywood/architecture/android/components/photocomponent/PhotoTakerActivity.java
package net.lonelywood.architecture.android.components.photocomponent;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.v7.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

public class PhotoTakerActivity extends AppCompatActivity {

    public static final String ExtraFileName = "fileName";

    private static List<PhotoTakerListener> listeners = new ArrayList<>();

    private String mFileName;
    private final int RequestImageCapture = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (savedInstanceState != null) {
            mFileName = savedInstanceState.getString(ExtraFileName);
        } else {
            mFileName = getIntent().getExtras().getString(ExtraFileName);
        }

        Uri uri = Uri.parse(mFileName);

        Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        intent.putExtra(MediaStore.EXTRA_OUTPUT, uri);

        if (intent.resolveActivity(getPackageManager()) != null) {
            startActivityForResult(intent, RequestImageCapture);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RequestImageCapture) {
            if (resultCode == RESULT_CANCELED){
                for (PhotoTakerListener listener: listeners) {
                    listener.onPhotoTaken(false, null);
                }
            } else if (resultCode == RESULT_OK){
                for (PhotoTakerListener listener: listeners){
                    listener.onPhotoTaken(true, mFileName);
                }
            }
        }

        finish();
    }

    public static void addListener(PhotoTakerListener listener){
        listeners.add(listener);
    }

    public static void removeListener(PhotoTakerListener listener) {
        listeners.remove(listener);
    }
}

File: Android-Kotlin/app/src/main/java/com/github/lonelywood/architecture/android/components/photocomponent/PhotoTakerActivity.kt
package com.github.lonelywood.architecture.android.components.photocomponent

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import androidx.appcompat.app.AppCompatActivity

interface PhotoTakerListener {
    fun onPhotoTaken(success: Boolean, path: String)
}

class PhotoTakerActivity: AppCompatActivity() {
    companion object {
        const val ExtraFileName: String = "fileName"

        private val listeners: ArrayList<PhotoTakerListener> = ArrayList()

        fun addListener(listener: PhotoTakerListener) { listeners.add(listener) }
        fun removeListener(listener: PhotoTakerListener) { listeners.remove(listener) }
    }

    private val REQUEST_IMAGE_CAPTURE = 1
    private var fileName: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState != null) {
            fileName = savedInstanceState.getString(ExtraFileName)
        } else {
            fileName = intent?.extras?.getString(ExtraFileName)
        }

        val uri: Uri = Uri.parse(fileName)

        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)

        grantUriPermissionsForIntent(intent, uri)

        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
        intent.putExtra(MediaStore.EXTRA_OUTPUT, uri)

        if (intent.resolveActivity(packageManager) != null) {
            startActivityForResult(intent, REQUEST_IMAGE_CAPTURE)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQUEST_IMAGE_CAPTURE) {
            if (resultCode == Activity.RESULT_CANCELED) {
                for (listener in listeners) {
                    listener.onPhotoTaken(false, "")
                }
            } else if (resultCode == Activity.RESULT_OK) {
                for (listener in listeners) {
                    listener.onPhotoTaken(true, fileName!!)
                }
            }
        }

        finish()
    }

    private fun grantUriPermissionsForIntent(intent: Intent, uri: Uri) {
        val resInfoList = applicationContext.packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)

        for (resInfo in resInfoList) {
            val packageName = resInfo.activityInfo.packageName
            applicationContext.grantUriPermission(packageName, uri, Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            applicationContext.grantUriPermission(packageName, uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }
}